geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/utils/NailgunServer.scala

```diff
@@ -0,0 +1,227 @@
+/***********************************************************************
+ * Copyright (c) 2013-2021 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.tools.utils
+
+import java.net.InetAddress
+import java.util.concurrent.atomic.{AtomicInteger, AtomicLong}
+import java.util.concurrent.{Executors, Future, Phaser, TimeUnit}
+
+import com.beust.jcommander.validators.PositiveInteger
+import com.beust.jcommander.{JCommander, Parameter}
+import com.codahale.metrics.{MetricRegistry, Timer}
+import com.facebook.nailgun.{NGConstants, NGServer}
+import com.github.benmanes.caffeine.cache.{CacheLoader, Caffeine, LoadingCache}
+import org.locationtech.geomesa.tools.Command
+import org.locationtech.geomesa.tools.utils.NailgunServer.{CommandStat, NailgunAware}
+import org.locationtech.geomesa.tools.utils.ParameterConverters.DurationConverter
+import org.locationtech.geomesa.utils.concurrent.ExitingExecutor.DaemonThreadFactory
+
+import scala.concurrent.duration.Duration
+
+class NailgunServer(addr: InetAddress, port: Int, sessionPoolSize: Int, timeoutMillis: Int, idleTimeoutMillis: Long)
+    extends NGServer(addr, port, sessionPoolSize, timeoutMillis) {
+
+  import scala.collection.JavaConverters._
+
+  private val registry = new MetricRegistry()
+  private val requests = new Phaser(1) // 1 for the idle timer
+
+  private val start = System.currentTimeMillis()
+  private val lastRequest = new AtomicLong(System.currentTimeMillis())
+
+  private val timers: LoadingCache[String, (AtomicInteger, Timer)] = Caffeine.newBuilder().build(
+    new CacheLoader[String, (AtomicInteger, Timer)]() {
+      override def load(k: String): (AtomicInteger, Timer) = (new AtomicInteger(0), registry.timer(k))
+    }
+  )
+
+  private val idle: Future[_] =
```

Comment:
Could we use a ScheduledExecutorService to check the state (e.g., stop when !executingSomething and lastCall is "old")?

The Thread.sleeps worry me for no good reason.

Reply:
i did initially implement this with a scheduled executor service and re-submitting the job each time, but i ran into issues with it dying appropriately when the nailgun was shut down explicitly. iirc, scheduled executors will keep the jvm up if they have a scheduled job still. hmm, although now that i'm looking at it the executor should probably be shut down after submitting the job... i'll test that out real quick. but having a daemon thread mostly sleeping in the background seemed acceptable. i suspect a scheduled executor service would be doing something similar with sleeping/idle threads internally.

Reply:
If you are right about the ScheduledExecutors.... I imagine there's either a way to set the Service to shutdown on JVM exit or we could try adding a shutdownhook to remove the scheduled task.

At least one of those options ought to work...

Reply:
here's the initial impl that tried to do pretty much that, but it didn't quite work correctly: https://github.com/locationtech/geomesa/blob/3c053e11f3079dbeb8751e5f8da7b222d7baca77/geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/utils/NailgunServer.scala#L32-L70

Reply:
from what i can tell the difference is `TIMED_WAITING (sleeping)` vs `TIMED_WAITING (parking)`, which doesn't really seem to matter... if there were multiple tasks or threads it would be better to use the executor, but for one thread and task it seems equivalent. the current approach seems simpler and clearer to me, as well, but maybe that's just me.

I'm nervous to change it since i had issues with the original scheduled executor and this seems to be working correctly.